### PR TITLE
feat(chat): direct chat flow

### PR DESCRIPTION
**Pull request by sieuhuflit**
## Overview

- [x] Add create chat button
- [x] Add dropdown to show 3 options when wanna create new chat
- [x] Implement search users matrix
- [x] Able to select/deselect users
- [x] Will create private group if choose multiple users
- [x] Will create direct chat if only choose 1 user 

## Demo

Sep-18-2023 13-04-46

### CODE REVIEW

**Review by mthinh — APPROVED** [review]

File: lib/core/application/chat/new_chat_bloc/new_chat_bloc.dart [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+    emit(state.copyWith(isSearching: true));
+    var text = event.text;
+    try {
+      final result = await matrixService.client.searchUserDirectory(
+        text!,
+        limit: 20,
+      );
+      emit(state.copyWith(isSearching: false));
+      emit(state.copyWith(userSearchResult: result));
+    } catch (e) {
```
Comment (roncoderpush) [inline lib/core/application/chat/new_chat_bloc/new_chat_bloc.dart:39]: Could you merge 2 state into 1?

File: lib/core/application/chat/new_chat_bloc/new_chat_bloc.dart [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+
+  _onDeselectUser(NewChatEventDeselectUser event, Emitter emit) {
+    final selectedUsers = List<Profile>.from(state.selectedUsers);
+    selectedUsers.remove(event.user);
+    emit(state.copyWith(selectedUsers: selectedUsers));
+  }
+
+  startChat(BuildContext context) async {
+    final List<String> selectedUserIds =
+        state.selectedUsers.map((profile) => profile.userId).toList();
```
Comment (roncoderpush) [inline lib/core/application/chat/new_chat_bloc/new_chat_bloc.dart:58]: Avoid using `context` inside BloC class, the purpose of the BloC is to separate logic and UI. This is violated the principles of BloC
Try emit an event back to UI and handle navigation at there using BlocListener

File: lib/core/presentation/pages/chat/chat_list/views/chat_list_page_view.dart [hunk 3]
```diff
@@ -127,6 +128,9 @@ class ChatListPageView extends StatelessWidget {
           ),
         ),
       ),
+      floatingActionButton: CreateChatButton(
+        onTap: () => null,
+      ),
```
Comment (roncoderpush) [inline lib/core/presentation/pages/chat/chat_list/views/chat_list_page_view.dart:133]: Should this navigate or do any logic here?

File: lib/core/presentation/pages/chat/new_chat/views/new_chat_view.dart [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+      onTap: onTap,
+      child: Container(
+        padding: EdgeInsets.symmetric(horizontal: 15.w),
+        alignment: Alignment.centerRight,
+        child: Text(
+          t.common.start,
+          style: TextStyle(
+            color: colorScheme.onSurface,
+            fontSize: 16.0, // Adjust the font size as needed
+          ),
```
Comment (roncoderpush) [inline lib/core/presentation/pages/chat/new_chat/views/new_chat_view.dart:143]: Should use `Typo` instead

File: lib/core/presentation/pages/chat/new_chat/widgets/search_user_input.dart [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+            onChanged: onChanged,
+            style: TextStyle(
+              fontWeight: FontWeight.w400,
+              color: colorScheme.onSurfaceVariant,
+            ),
+            decoration: InputDecoration(
+              hintText: t.chat.searchPeople,
+              hintStyle: TextStyle(
+                color: colorScheme.onSurfaceVariant,
+              ),
```
Comment (roncoderpush) [inline lib/core/presentation/pages/chat/new_chat/widgets/search_user_input.dart:65]: Undefined text size, use `Typo` instead

File: lib/core/presentation/pages/chat/new_chat/widgets/search_user_input.dart [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+              ],
+            ),
+          ),
+          TextField(
+            cursorColor: colorScheme.onSurfaceVariant,
+            onChanged: onChanged,
+            style: TextStyle(
+              fontWeight: FontWeight.w400,
+              color: colorScheme.onSurfaceVariant,
+            ),
```
Comment (roncoderpush) [inline lib/core/presentation/pages/chat/new_chat/widgets/search_user_input.dart:60]: Undefined text size, use `Typo`

File: lib/core/presentation/pages/chat/new_chat/widgets/search_user_item.dart [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+          horizontal: Spacing.small,
+        ),
+        child: Row(
+          children: [
+            buildAvatar(),
+            SizedBox(width: Spacing.xSmall),
+            Expanded(
+              child: Column(
+                crossAxisAlignment: CrossAxisAlignment.start,
+                children: [
```
Comment (roncoderpush) [inline lib/core/presentation/pages/chat/new_chat/widgets/search_user_item.dart:50]: Remove unnecessary Column

File: lib/core/presentation/widgets/chat/create_chat_button.dart [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+    this.onTap,
+  });
+  final Function()? onTap;
+
+  buildCustomButton() {
+    return Container(
+      width: 54.h,
+      height: 54.h,
+      padding: const EdgeInsets.all(15),
+      decoration: const BoxDecoration(
```
Comment (roncoderpush) [inline lib/core/presentation/widgets/chat/create_chat_button.dart:21]: add `.w` to make it responsive

**Review by roncoderpush — CHANGES_REQUESTED** [review]

File: lib/core/presentation/pages/chat/chat_list/views/chat_list_page_view.dart (hunk 3 above)
Comment (sieuhuflit) [reply lib/core/presentation/pages/chat/chat_list/views/chat_list_page_view.dart:133]: Ah, redundant props on my side :v thanks 👍

File: lib/core/application/chat/new_chat_bloc/new_chat_bloc.dart (hunk 2 above)
Comment (sieuhuflit) [reply lib/core/application/chat/new_chat_bloc/new_chat_bloc.dart:58]: Thanks 👍

**Review by roncoderpush — APPROVED** [review]

**Pull request merged**